azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,277 @@
+                else: 
+                    id["numModules"] = len(module_objs)
+
+                # primary and secondary keys show up in the "show" output but not in the "list" output
+                if id["authenticationType"] == "sas":
+                    id2 = _iot_device_show(self.target, id["deviceId"])
+                    id["symmetricKey"] = id2["authentication"]["symmetricKey"]
+
+                twin = _iot_device_twin_show(self.target, id["deviceId"])
+
+                json.dump(id, f)
+                f.write('\n')
+                json.dump(twin, f)
+                f.write('\n')
+
+                for module in module_objs:
+                    module = vars(module)
+
+                    if(module["module_id"] not in ["$edgeAgent", "$edgeHub"]):
+
+                        module.pop('connection_state_updated_time')
+                        module.pop('last_activity_time')
+                        module.pop('authentication')
+
+                        module_twin = _iot_device_module_twin_show(self.target, id["deviceId"], module["module_id"])
+
+                        module["authenticationType"] = module_twin["authenticationType"]
+                        if (module["authenticationType"] == "sas"):
+                            module2 = _iot_device_module_show(self.target, id["deviceId"], module["module_id"])
+                            module["symmetricKey"] = module2["authentication"]["symmetricKey"]
+
+                        if (module["authenticationType"] == "certificateAuthority"):
+                            module2 = _iot_device_module_show(self.target, id["deviceId"], module["module_id"])
+                            module["x509Thumbprint"] = module2["authentication"]["x509Thumbprint"]
+
+                        json.dump(module, f)
+                        f.write('\n')
+                        json.dump(module_twin, f)
+                        f.write('\n')
+
+        logger.info("Saved state of IoT Hub '{}' to {}".format(self.hub_name, filename))
+
+    def upload_device_identity(self, identity):
+        device_id = identity["deviceId"]
+        auth_type = identity["authenticationType"]
+        edge = identity["capabilities"]["iotEdge"]
+        status = identity["status"]
+        ptp = identity["x509Thumbprint"]["primaryThumbprint"]
+        stp = identity["x509Thumbprint"]["secondaryThumbprint"]
+
+        if "status_reason" in identity.keys():
+            status_reason = identity["statusReason"]
+        else:
+            status_reason = None
+
+        if(auth_type == "sas"):
+            pk = identity["symmetricKey"]["primaryKey"]
+            sk = identity["symmetricKey"]["secondaryKey"]
+
+            iot_device_create(self.cmd, device_id, self.hub_name, edge, primary_key=pk, secondary_key=sk, status=status, \
+                status_reason=status_reason, resource_group_name=self.rg, auth_type_dataplane=self.auth_type)
+
+        elif(auth_type == "selfSigned"):
```

Comment:
Please use the **DeviceAuthType** Enum defined in **azext_iot\common\shared.py** for the value of _auth_method_ parameter.